zerver/tests/test_cache.py

```diff
@@ -163,6 +182,33 @@ def get_user_function_can_return_none(user_id: int) -> Optional[UserProfile]:
         self.assertEqual(result_two, None)
         self.assert_length(queries, 0)
 
+    def test_cache_with_key_database(self) -> None:
+        def good_cache_key_function(user_id: int) -> str:
+            return f"CacheWithKeyDecoratorTest:good_cache_key:{user_id}"
+
+        @cache_with_key(good_cache_key_function, cache_name="database", with_statsd_key="stk")
+        def get_user_function_with_good_cache_keys(user_id: int) -> UserProfile:
+            return UserProfile.objects.get(id=user_id)
+
+        hamlet = self.example_user("hamlet")
+        mock_backend = get_cache_backend(None)
+
+        with queries_captured() as queries:
+            with patch("zerver.lib.cache.get_cache_backend", return_value=mock_backend):
+                result = get_user_function_with_good_cache_keys(hamlet.id)
+
+        self.assertEqual(result, hamlet)
+        self.assert_length(queries, 1)
+
+        # The previous function call should have cached the result correctly, so now
+        # no database queries should happen:
+        with queries_captured(keep_cache_warm=True) as queries_two:
+            with patch("zerver.lib.cache.get_cache_backend", return_value=mock_backend):
+                result_two = get_user_function_with_good_cache_keys(hamlet.id)
```

Comment:
I don't understand this test -- doesn't the `database` cache need to do a database query? I think it might be better to skip `mock_backend` and just check that we do 1 database query rather than the 3 that would occur on a cache miss (One to check the cache, one to get the data, and one to store into the cache).